README.md
xbfuse is a program that allows you to mount the filesystem of a 
Microsoft Xbox or Xbox 360 DVD as a read-only part of the Linux 
filesystem. This allows the user to browse the directory structure and 
read the files within. 

xbfuse is made possible with Filesystem in Userspace (FUSE), available 
at:

  http://fuse.sourceforge.net/

Note that it is not usually possible to simply read an Xbox/360 DVD in
an ordinary computer's DVD-ROM drive. The optical drives in Microsoft 
consoles have special firmware which allows them to access areas of the 
disc that are effectively invisible to most DVD-ROM drives. In order to 
mount a filesystem, generally, you will have to rip the proper 
sector image from the disc using special hardware and tools, or contact 
another source who has already done so.

Note also that there are likely to be bugs and perhaps even security 
problems. xbfuse is currently meant as primarily an experimental 
research tool for studying Microsoft Xbox/360 discs.


Requirements:
	- Linux 2.4.x or 2.6.x (as of 2.6.14 FUSE is part of the
	  kernel, but you still need user libraries)
	- FUSE (http://fuse.sourceforge.net) 2.5.x or higher

Build:
	./configure && make

Install:
	make install

Usage:
	xbfuse <image_file> <mount_point>

	To unmount previously mounted file, use:
	fusermount -u <mount_point>

	To debug, or investigate how xbfuse examines the filesystem:
	xbfuse <image_file> <mount_point> -d


NEWS
release 1.0
	* initial release


src/main.c
/*
 *  Copyright (C) 2006 Mike Melanson (mike at multimedia.cx)
 *    based on code by:
 *  Copyright (C) 2005 Janusz Dziemidowicz (rraptorr@nails.eu.org)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*!
 * \file main.c
 * \author Mike Melanson
 * \brief Main program function.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

//! FUSE library compliance level.
#define FUSE_USE_VERSION 25
#include <fuse.h>

#include "xdvdfs.h"

/*!
 * \brief Flag indicating wheter we should run in quiet mode (0 - no,
 * all other values - yes).
 */
int quiet;

/*!
 * \brief Main function.
 */
int main(int argc, char *argv[])
{
	char **nargv;
	int nargc, i, j;

	if (argc < 3) {
		fprintf
		    (stderr,
		     "Usage: %s <archive_file> <mount_point> [<options>] [<FUSE library options>]\n\n",
		     argv[0]);
		fprintf(stderr, "Available options:\n");
		fprintf(stderr,
			"\t-q - quiet mode (print only error messages)\n");
		exit(EXIT_FAILURE);
	}

	for (i = 1; i < argc; i++) {
		if (!strcmp(argv[i], "-q")) {
			quiet = 1;
			for (j = i; j < argc - 1; j++)
				argv[j] = argv[j + 1];
			argc--;
			break;
		}
	}

	nargc = argc - 1;
	nargv = (char **)malloc(nargc * sizeof(char *));

	// skip the resource filename
	nargv[0] = argv[0];
	for (i = 1; i < nargc; i++)
		nargv[i] = argv[i + 1];

	// try to open the file
	xbfs_fd = open(argv[1], O_RDONLY);
	if (xbfs_fd < 0) {
		perror(argv[1]);
		exit(EXIT_FAILURE);
	}

	return fuse_main(nargc, nargv, &xbfs_operations);
}


src/tree.c
/*
 *  Copyright (C) 2005 Janusz Dziemidowicz (rraptorr@nails.eu.org)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*!
 * \file tree.c
 * \author Janusz Dziemidowicz
 * \brief Directory hierarchy abstraction file.
 */

#include "tree.h"

void tree_insert(struct tree *root, const char *path, int length,
		 off_t offset, long size)
{
	char *pos;
	struct tree *node;

	if (!path || !*path)
		return;

	if ((pos = memchr(path, '/', length))) {
		// Path contains directory.
		*pos = '\0';
		node = root->sub;

		// Check if this directory was already inserted by earlier calls.
		while (node) {
			if (!strcmp(node->name, path)) {
				tree_insert(node, pos + 1,
					    length - (pos + 1 - path),
					    offset, size);
				return;
			}
			node = node->next;
		}

		// Create new directory.
		node = (struct tree *)malloc(sizeof(struct tree));
		node->name = strdup(path);
		node->is_dir = 1;
		node->offset = 0;
		node->size = 0;
		node->nsubdirs = 0;
		node->sub = NULL;
		node->next = root->sub;

		// Connect new directory under current directory.
		root->sub = node;
		root->nsubdirs++;

		// Insert remaining parts of path in new directory.
		tree_insert(node, pos + 1, length - (pos + 1 - path),
			    offset, size);
	} else {
		// No more directories in path. Just create new file
		// under current directory.
		node = (struct tree *)malloc(sizeof(struct tree));
		node->name = strndup(path, length);
		node->is_dir = 0;
		node->offset = offset;
		node->size = size;
		node->nsubdirs = 0;
		node->sub = NULL;
		node->next = root->sub;
		root->sub = node;
	}
}

struct tree *tree_find_entry(struct tree *root, const char *path)
{
	struct tree *node, *ret;
	const char *next;

	if (!root)
		return NULL;

	if (!strcmp(path, root->name))
		return root;

	if (strncmp(path, root->name, strlen(root->name)))
		return NULL;

	next = path + strlen(root->name);
	if (*next != '/')
		return NULL;
	next++;

	if (!*next)
		return root;

	node = root->sub;
	while (node) {
		ret = tree_find_entry(node, next);
		if (ret)
			return ret;

		node = node->next;
	}

	return NULL;
}

void tree_free(struct tree *root)
{
	struct tree *node, *next;

	node = root->sub;
	while (node) {
		next = node->next;
		tree_free(node);
		node = next;
	}

	free(root->name);
	free(root);
}

struct tree *tree_empty(void)
{
	struct tree *ret;

	ret = (struct tree *)malloc(sizeof(struct tree));
	ret->name = strdup("");
	ret->is_dir = 1;
	ret->offset = 0;
	ret->size = 0;
	ret->nsubdirs = 0;
	ret->sub = NULL;
	ret->next = NULL;

	return ret;
}

int tree_getattr(const char *path, struct stat *stbuf, struct tree *root,
		 int fd)
{
	struct tree *node;
	struct stat st;

	node = tree_find_entry(root, path);
	if (node) {
		memset(stbuf, 0, sizeof(struct stat));
		// Set UID and GID to current user
		stbuf->st_uid = getuid();
		stbuf->st_gid = getgid();
		if (node->is_dir) {
			stbuf->st_mode =
			    S_IFDIR | S_IRUSR | S_IRGRP | S_IROTH | S_IXUSR
			    | S_IXGRP | S_IXOTH;
			// Directory should have number of links set
			// to 2 + number of subdirectories (not
			// files), this makes find work.
			stbuf->st_nlink = 2 + node->nsubdirs;
		} else {
			stbuf->st_mode = S_IFREG | S_IRUSR | S_IRGRP | S_IROTH;
			stbuf->st_nlink = 1;
			stbuf->st_size = node->size;
		}

		// Set all times to the values from PACK file.
		fstat(fd, &st);
		stbuf->st_atime = st.st_atime;
		stbuf->st_mtime = st.st_mtime;
		stbuf->st_ctime = st.st_ctime;
		return 0;
	} else
		return -ENOENT;
}

int tree_open(const char *path, struct fuse_file_info *fi, struct tree *root)
{
	struct tree *node;

	if (fi->flags & O_WRONLY)
		return -EROFS;

	node = tree_find_entry(root, path);

	if (!node)
		return -ENOENT;

	return 0;
}

int tree_read(const char *path, char *buf, size_t size,
	      off_t offset, struct fuse_file_info *fi, struct tree *root,
	      int fd, pthread_mutex_t * mutex)
{
	struct tree *node;
	int ret;

	node = tree_find_entry(root, path);
	if (!node)
		return -ENOENT;

	if (node->is_dir)
		return -EISDIR;

	if (offset >= node->size)
		return 0;
	if (offset + size > node->size)
		size = node->size - offset;

	// This is needed to ensure, that no other thread will change
	// seek point before we call read().
	pthread_mutex_lock(mutex);

	lseek(fd, node->offset + offset, SEEK_SET);
	ret = read(fd, buf, size);

	pthread_mutex_unlock(mutex);

	return ret;
}

int tree_opendir(const char *path, struct fuse_file_info *fi, struct tree *root)
{
	struct tree *node = tree_find_entry(root, path);

	if (!node)
		return -ENOENT;

	if (!node->is_dir)
		return -ENOTDIR;

	return 0;
}

int tree_readdir(const char *path, void *buf,
		 fuse_fill_dir_t filler, off_t offset,
		 struct fuse_file_info *fi, struct tree *root)
{
	struct tree *node;

	node = tree_find_entry(root, path);

	if (!node)
		return -ENOENT;

	if (!node->is_dir)
		return -ENOTDIR;

	filler(buf, ".", NULL, 0);
	filler(buf, "..", NULL, 0);

	node = node->sub;
	while (node) {
		filler(buf, node->name, NULL, 0);
		node = node->next;
	}

	return 0;
}


src/tree.h
/*
 *  Copyright (C) 2005 Janusz Dziemidowicz (rraptorr@nails.eu.org)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*!
 * \file tree.h
 * \author Janusz Dziemidowicz
 * \brief Directory hierarchy abstraction header file.
 */

#ifndef _TREE_H_
#define _TREE_H_

//! This is needed for strndup.
#define _GNU_SOURCE

#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <pthread.h>

//! FUSE library compliance level.
#define FUSE_USE_VERSION 25
#include <fuse.h>

/*!
 * \brief GRAF directory hierarchy structure.
 *
 * This structure represents directory structure from GRAF. You should
 * insert files into this structure (filenames can contain '/',
 * corresponding directories will be made automatically). There are
 * also FUSE operations on this structure provided, which can be used
 * to quickly implement FUSE filesystem.
 */
struct tree {
	/*!
	 * \brief Filename.
	 *
	 * Should be empty string in the root of the tree. Otherwise
	 * searching function won't work.
	 */
	char *name;

	/*!
	 * \brief Flag indicating that this a directory.
	 */
	char is_dir;

	/*!
	 * \brief Offset of the file inside of GRAF.
	 */
	off_t offset;

	/*!
	 * \brief Size of the file inside of GRAF.
	 */
	off_t size;

	/*!
	 * \brief Number of subdirectories.
	 *
	 * This is used to calculate correct number of hard links,
	 * which for directories should be 2 + number of
	 * subdirectories. This is especially needed to make find
	 * work. It will be incremented when inserting directories.
	 */
	int nsubdirs;

	/*!
	 * \brief Subdirectories.
	 *
	 * If the current file is a regular file this field is NULL. Otherwise
	 * it is a pointer to first element of directory.
	 */
	struct tree *sub;

	/*!
	 * \brief Next file.
	 *
	 * This is a pointer to the next file on the same directory
	 * level.
	 */
	struct tree *next;
};

/*!
 * \brief Insert path into representation of GRAF firectory tree.
 *
 * This function inserts given pathname into the \c tree structure,
 * which describes directory structure inside of GRAF. Every needed
 * subdirectories will be created by this function.
 *
 * \param root \c tree root, at which \c path should be inserted.
 * \param path path that should be inserted. This should be relative
 * name (no '/' in the beginning). Please note that it will be
 * destroyed.
 * \param length length of \c path (as sometimes it is not zero
 * terminated).
 * \param offset offset (from \c packentry) that should be stored in
 * the tree node corresponding to given \c path.
 * \param size size (from \c packentry) that should be stored in the
 * tree node correspoding to given \c path.
 */
void tree_insert(struct tree *root, const char *path, int length,
		 off_t offset, long size);

/*!
 * \brief Find given path in GRAF directory structure.
 *
 * \param root \c tree root, at which searching should be started.
 * \param path path to find, it should be exactly as provided by FUSE
 * (so it should start with '/').
 * \return pointer to node containing information about given \c path
 * or NULL if not found.
 */
struct tree *tree_find_entry(struct tree *root, const char *path);

/*!
 * \brief Free entire GRAF directory hierarchy structure.
 *
 * \param root root of the directory \c tree to be freed.
 */
void tree_free(struct tree *root);

/*!
 * \brief Create empty directory structure.
 *
 * This functions creates empty directory structures with all fields
 * set as needed.
 */
struct tree *tree_empty(void);

/*!
 * \brief FUSE getattr operation.
 *
 * This is FUSE compatible getattr operation which gets file
 * information from \c tree structure.
 * \param path file path.
 * \param stbuf stats will be stored here.
 * \param root tree root.
 * \param fd data file descriptor.
 * \return 0 on success, -errno otherwise.
 */
int tree_getattr(const char *path, struct stat *stbuf, struct tree *root,
		 int fd);

/*!
 * \brief FUSE open operation.
 *
 * This is FUSE compatible open operation which gets file
 * information from \c tree structure.
 * \param path file path.
 * \param fi FUSE file information.
 * \param root tree root.
 * \return 0 on success, -errno otherwise.
 */
int tree_open(const char *path, struct fuse_file_info *fi, struct tree *root);

/*!
 * \brief FUSE read operation.
 *
 * This is FUSE compatible read operation which gets file
 * information from \c tree structure.
 * \param path file path.
 * \param buf read buffer.
 * \param size size of \c buf.
 * \param offset read offset.
 * \param fi FUSE file information.
 * \param root tree root.
 * \param fd data file descriptor.
 * \param mutex file operations mutex related to \c fd.
 * \return number of bytes read on success, -errno otherwise.
 */
int tree_read(const char *path, char *buf, size_t size,
	      off_t offset, struct fuse_file_info *fi, struct tree *root,
	      int fd, pthread_mutex_t * mutex);

/*!
 * \brief FUSE opendir operation.
 *
 * This is FUSE compatible opendir operation which gets file
 * information from \c tree structure.
 * \param path directory path.
 * \param fi FUSE file information.
 * \param root tree root.
 * \return 0 on success, -errno otherwise.
 */
int tree_opendir(const char *path, struct fuse_file_info *fi,
		 struct tree *root);

/*!
 * \brief FUSE readdir operation.
 *
 * This is FUSE compatible readdir operation which gets file
 * information from \c tree structure.
 * \param path directory path.
 * \param buf output buffer.
 * \param filler filler function.
 * \param offset output offset.
 * \param fi FUSE file information.
 * \param root tree root.
 * \return 0 on success, -errno otherwise.
 */
int tree_readdir(const char *path, void *buf,
		 fuse_fill_dir_t filler, off_t offset,
		 struct fuse_file_info *fi, struct tree *root);

#endif				// _TREE_H_


src/xdvdfs.h
/*
 *  Copyright (C) 2006 Mike Melanson (mike at multimedia.cx)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*!
 * \file xbfs.h
 * \author Mike Melanson
 * \brief XBFS file support header file.
 */

#ifndef _XBFS_H_
#define _XBFS_H_

#define _GNU_SOURCE

#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>

#include <pthread.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

/*!
 * \brief Basic information about one XBFS file.

 * This structure contains basic information about one XBFS file,
 * including information needed by FUSE system. It is stored as \c
 * private_data of FUSE context.
 */
struct xbfsfile {
	/*!
	 * \brief XBFS file descriptor.
	 */
	int fd;

	/*!
	 * \brief Disk operations mutex.
	 *
	 * This mutex is used to synchronize disk operations on XBFS
	 * file.  Whenever one need to read/write/lseek XBFS file, this
	 * mutex should be locked with \c pthread_mutex_lock(). After
	 * operation has finished (also after an error), this mutex
	 * should be unlocked with \c pthread_mutex_unlock().
	 */
	pthread_mutex_t mutex;

	/*!
	 * \brief Size of the XBFS file.
	 */
	off_t size;

	/*!
	 * \brief Directory tree of the XBFS file.
	 *
	 * This field contains directory tree of the XBFS file (which
	 * doesn't constain any directories, only files). This is
	 * filled by \c xbfs_init().
	 */
	struct tree *tree;
};

extern struct fuse_operations xbfs_operations;

extern int xbfs_fd;

//! Treat given memory address as a 16-bit big-endian integer.
#define BE_16(x)  ((((uint8_t*)(x))[0] << 8) | ((uint8_t*)(x))[1])

//! Treat given memory address as a 32-bit big-endian integer.
#define BE_32(x)  ((((uint8_t*)(x))[0] << 24) | \
                   (((uint8_t*)(x))[1] << 16) | \
                   (((uint8_t*)(x))[2] << 8) | \
                    ((uint8_t*)(x))[3])

//! Treat given memory address as a 16-bit little-endian integer.
#define LE_16(x)  ((((uint8_t*)(x))[1] << 8) | ((uint8_t*)(x))[0])

//! Treat given memory address as a 32-bit little-endian integer.
#define LE_32(x)  ((((uint8_t*)(x))[3] << 24) | \
                   (((uint8_t*)(x))[2] << 16) | \
                   (((uint8_t*)(x))[1] << 8) | \
                    ((uint8_t*)(x))[0])

#endif				// _XBFS_H_


src/xdvdfs.c
/*
 *  Copyright (C) 2006 Mike Melanson (mike at multimedia.cx)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*!
 * \file xbfs.c
 * \author Mike Melanson
 * \brief Interpret the Xbox XDVD filesystem
 */

#include <ctype.h>

#include "tree.h"
#include "xdvdfs.h"

// cast this constant as an unsigned long long in the hopes that the compiler will
// always to the right 64-bit math
#define SECTOR_SIZE 2048ULL
#define NAME_MAX_SIZE 1024
#define XDVD_SIGNATURE "MICROSOFT*XBOX*MEDIA"
#define XDVD_SIGNATURE_SIZE 0x14

// global file description since main program needs to access it
int xbfs_fd;

//! Extract \c xbfsfile structure from FUSE context.
static inline struct xbfsfile *get_xbfsfile_from_context(void)
{
	return (struct xbfsfile *)fuse_get_context()->private_data;
}

// **********************************************************************
// xbfs operations
// **********************************************************************

// **********************************************************************
// FUSE operations
// Here comes FUSE operations, please consult fuse.h for description of
// each operation.
// **********************************************************************

/*!
 * \brief Get file attributes.
 */
static int xbfs_getattr(const char *path, struct stat *stbuf)
{
	return tree_getattr(path, stbuf, get_xbfsfile_from_context()->tree,
		get_xbfsfile_from_context()->fd);
}

/*!
 * \brief File open operation.
 */
static int xbfs_open(const char *path, struct fuse_file_info *fi)
{
	return tree_open(path, fi, get_xbfsfile_from_context()->tree);
}

/*!
 * \brief Read data from an open file.
 */
static int xbfs_read(const char *path, char *buf, size_t size,
		    off_t offset, struct fuse_file_info *fi)
{
	return tree_read(path, buf, size, offset, fi,
		 get_xbfsfile_from_context()->tree,
		 get_xbfsfile_from_context()->fd,
		 &get_xbfsfile_from_context()->mutex);
}

/*!
 * Open directory.
 */
static int xbfs_opendir(const char *path, struct fuse_file_info *fi)
{
	return tree_opendir(path, fi, get_xbfsfile_from_context()->tree);
}

/*!
 * \brief Read directory.
 */
static int xbfs_readdir(const char *path, void *buf,
		       fuse_fill_dir_t filler, off_t offset,
		       struct fuse_file_info *fi)
{
	return tree_readdir(path, buf, filler, offset, fi,
			    get_xbfsfile_from_context()->tree);
}

// xbfs_recurse_file_subtree() calls this function
static void xbfs_recurse_directory(
	char *name_buffer,
	int fd,
	off_t filesystem_base_offset,
	unsigned int dir_entry_sector,
	unsigned int dir_entry_size,
	struct tree *root);

/*!
 * \brief Recurse through a directory structure.
 */
static void xbfs_recurse_file_subtree(
	char *name_buffer,
	int fd,
	off_t filesystem_base_offset,
	unsigned char *dir_entry,
	unsigned int dir_entry_size,
	int filerecord_offset,
	struct tree *root)
{
	unsigned int subtree_offset;
	unsigned int file_sector;
	unsigned int file_size;
	unsigned char file_attributes;
	unsigned char filename_size;
	int is_dir;
	int start_index;
	char name_copy[NAME_MAX_SIZE];

	// if there is not enough data left in the buffer for a minimal file record, get out
	if (filerecord_offset + 0xD >= dir_entry_size)
		return;

	// process left subtree
	subtree_offset = LE_16(&dir_entry[filerecord_offset]) * 4;
	if (subtree_offset)
		xbfs_recurse_file_subtree(name_buffer, fd,
			filesystem_base_offset, dir_entry, dir_entry_size, 
			subtree_offset, root);

	// process file
	file_sector = LE_32(&dir_entry[filerecord_offset + 4]);
	file_size = LE_32(&dir_entry[filerecord_offset + 8]);
	file_attributes = dir_entry[filerecord_offset + 0xC];
	is_dir = file_attributes & 0x10;
	filename_size = dir_entry[filerecord_offset + 0xD];
	strcpy(name_copy, name_buffer);
	start_index = strlen(name_copy);
	memcpy(&name_copy[start_index], &dir_entry[filerecord_offset + 0xE],
		filename_size);
	name_copy[start_index + filename_size] = '\0';

	if (is_dir) {
		name_copy[start_index + filename_size] = '/';
		name_copy[start_index + filename_size + 1] = '\0';
		xbfs_recurse_directory(name_copy, 
			fd,
			filesystem_base_offset,
			file_sector,
			file_size,
			root);
	} else {
		fprintf(stderr, " inserting %s: sector 0x%X, 0x%X bytes, attribute byte = 0x%X, %s\n", 
			name_copy, file_sector, file_size, file_attributes, 
			is_dir ? "directory" : "");

		tree_insert(root, name_copy, strlen(name_copy),
			filesystem_base_offset + file_sector * SECTOR_SIZE, file_size);
	}

	// process right subtree
	subtree_offset = LE_16(&dir_entry[filerecord_offset + 2]) * 4;
	if (subtree_offset)
		xbfs_recurse_file_subtree(name_buffer, fd,
			filesystem_base_offset, dir_entry, dir_entry_size, 
			subtree_offset, root);
}

/*!
 * \brief Recurse through a directory structure.
 */
static void xbfs_recurse_directory(
	char *name_buffer,
	int fd,
	off_t filesystem_base_offset,
	unsigned int dir_entry_sector,
	unsigned int dir_entry_size,
	struct tree *root)
{
	unsigned char *dir_entry;
	off_t current_offset = filesystem_base_offset;

	fprintf(stderr, "loading directory %s @ sector 0x%X, 0x%X bytes\n",
		(name_buffer[0]) ? name_buffer : "(root)",
		dir_entry_sector,
		dir_entry_size);

	// allocate a buffer and load the entire directory entry; round up to the
	// nearest sector
	dir_entry = (unsigned char *)malloc(dir_entry_size);
	if (!dir_entry)
		return;

	current_offset += dir_entry_sector * SECTOR_SIZE;
	lseek(fd, current_offset, SEEK_SET);
	if (read(fd, dir_entry, dir_entry_size) != dir_entry_size) {
		free(dir_entry);
		return;
	}

	xbfs_recurse_file_subtree(name_buffer, fd,
		filesystem_base_offset, dir_entry, dir_entry_size, 
		0, root);

	free(dir_entry);
}

/*!
 * \brief Initialize filesystem.
 */
static void *xbfs_init(void)
{
	char name_buffer[NAME_MAX_SIZE];
	unsigned int root_directory_sector;
	unsigned int root_directory_size;
	unsigned char timestamp[8];
	unsigned char sector_buffer[SECTOR_SIZE];
	off_t filesystem_base_offset = 0;

	struct xbfsfile *xbfs = (struct xbfsfile *)malloc(sizeof(struct xbfsfile));
	if (!xbfs) {
		fprintf(stderr,"not enough memory\n");
		close(xbfs_fd);
		fuse_exit(fuse_get_context()->fuse);
		return NULL;
	}

	xbfs->fd = xbfs_fd;

	// scan sectors until the signature is found
	lseek(xbfs->fd, filesystem_base_offset, SEEK_SET);
	while (1) {
		if (read(xbfs->fd, sector_buffer, SECTOR_SIZE) != SECTOR_SIZE) {
			fprintf(stderr, "XDVD signature (%s) not found\n", XDVD_SIGNATURE);
			close(xbfs->fd);
			free(xbfs);
			fuse_exit(fuse_get_context()->fuse);
			return NULL;
		}
		if (!strncmp((char *)sector_buffer, XDVD_SIGNATURE, XDVD_SIGNATURE_SIZE)) {
			// the actual root of the filesystem is 32 sectors back
			filesystem_base_offset -= (32 * SECTOR_SIZE);

			// process the volume descriptor
			root_directory_sector = LE_32(&sector_buffer[0x14]);
			root_directory_size = LE_32(&sector_buffer[0x18]);
			memcpy(timestamp, &sector_buffer[0x1C], 8);

			fprintf(stderr, "root @ sector 0x%X, 0x%X bytes; time = %02X %02X %02X %02X %02X %02X %02X %02X\n",
				root_directory_sector, root_directory_size,
				timestamp[0], timestamp[1], timestamp[2], timestamp[3],
				timestamp[4], timestamp[5], timestamp[6], timestamp[7]);

			break;
		}
		filesystem_base_offset += SECTOR_SIZE;
	}

	pthread_mutex_init(&xbfs->mutex, NULL);

	xbfs->tree = tree_empty();

	name_buffer[0] = 0;

	// build the tree
	xbfs_recurse_directory(name_buffer, xbfs->fd, filesystem_base_offset,
		root_directory_sector, root_directory_size, xbfs->tree);

	return (void *)xbfs;
}

/*!
 * \brief Clean up filesystem.
 */
static void xbfs_destroy(void *context)
{
	struct xbfsfile *xbfs = (struct xbfsfile *)context;

	if (xbfs) {
		close(xbfs->fd);
		tree_free(xbfs->tree);
		free(xbfs);
	}
}

/*!
 * \brief The FUSE file system operations.
 */
struct fuse_operations xbfs_operations = {
	.getattr = xbfs_getattr,
	.open = xbfs_open,
	.read = xbfs_read,
	.opendir = xbfs_opendir,
	.readdir = xbfs_readdir,
	.init = xbfs_init,
	.destroy = xbfs_destroy,
};
